cmd/oklog/ingeststore.go

```diff
@@ -445,22 +453,93 @@ func runIngestStore(args []string) error {
 			c.Stop()
 		})
 	}
-	{
-		c := store.NewCompacter(
-			storeLog,
+
+	topicLogs := store.NewTopicLogs(func(t string) (store.Log, error) {
+		return store.NewFileLog(
+			fsys,
+			filepath.Join(topicDir, t),
 			*segmentTargetSize,
-			*segmentRetain,
-			*segmentPurge,
-			compactDuration,
-			trashedSegments,
-			purgedSegments,
-			store.LogReporter{Logger: log.With(logger, "component", "Compacter")},
+			*segmentBufferSize,
+			store.LogReporter{Logger: logger},
 		)
+	})
+	// Initialize logs for existing topics from disk.
+	topics, err := fsys.ReadDir(topicDir)
+	if err != nil {
+		return err
+	}
+	for _, t := range topics {
+		if t.IsDir() {
+			if _, err = topicLogs.GetOrCreate(t.Name()); err != nil {
+				return err
+			}
+		}
+	}
+
+	{
+		newSegment := func(t string) (store.WriteSegment, error) {
+			log, err := topicLogs.GetOrCreate(t)
+			if err != nil {
+				return nil, err
+			}
+			return log.Create()
+		}
+		// TODO(fabxc): track delay between staging and demux completion in a metric.
+		demux := store.NewDemuxer(stagingLog.Oldest, newSegment)
+		stopc := make(chan struct{})
+		tickr := time.NewTicker(time.Second)
+
 		g.Add(func() error {
-			c.Run()
-			return nil
+			for {
+				select {
+				case <-stopc:
+					return nil
+				case <-tickr.C:
+					if err := demux.Next(); err != nil {
+						level.Error(logger).Log("msg", "demux failed", "err", err)
+					}
+				}
+			}
 		}, func(error) {
-			c.Stop()
+			close(stopc)
+			tickr.Stop()
+		})
+	}
+	{
+		tc := store.TopicCompacters{}
+		stopc := make(chan struct{})
+		tickr := time.NewTicker(time.Second)
+
+		g.Add(func() error {
+			for {
+				select {
+				case <-tickr.C:
+					// Ensure we have a compactor for each topic.
+					for t, l := range topicLogs.Clone() {
+						tc.Ensure(t, func() *store.Compacter {
+							return store.NewCompacter(
+								l,
+								*segmentTargetSize,
+								*segmentRetain,
+								*segmentPurge,
+								compactDuration,
+								trashedSegments,
+								purgedSegments,
+								store.LogReporter{Logger: log.With(logger, "component", "Compacter", "topic", t)},
+							)
+						})
+					}
+					// Run each compacters' next stage.
+					for _, c := range tc {
+						c.Next()
+					}
```

Comment:
Using Run/Stop is not really feasible here I think as we may end up with 1000 compactors competing for IO. Some parallelism might be okay though if this cannot keep up.

I generally prefer handling lifecycle entirely in main over blackbox-y Run/Stop methods – which works especially great with `run.Group`.